### PR TITLE
Adding SIBRA extension to SCION.

### PR DESCRIPTION
**Pull request by kormat**
This adds the code required to parse and pack the SIBRA extension
(including the SIBRA Lite modification), but doesn't include any code to
actually process the extension yet.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/netsec-ethz/scion/pull/558%23issuecomment-163319295%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23issuecomment-163320361%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23issuecomment-163632080%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47334357%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47334515%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47334940%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47335024%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23issuecomment-163886924%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47474335%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47474483%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47474780%22%2C%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23issuecomment-164388101%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23issuecomment-163319295%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%40pszalach%20%22%2C%20%22created_at%22%3A%20%222015-12-09T16%3A43%3A40Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40reischuk%20%22%2C%20%22created_at%22%3A%20%222015-12-09T16%3A47%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ready%20for%20review%20now.%22%2C%20%22created_at%22%3A%20%222015-12-10T14%3A17%3A09Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%2C%20%7B%22body%22%3A%20%22the%20rest%20lgtm%22%2C%20%22created_at%22%3A%20%222015-12-11T09%3A35%3A53Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22PR%20updated.%22%2C%20%22created_at%22%3A%20%222015-12-14T09%3A28%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20aba8ff163866ec86db5bc129068de5a9136dd944%20infrastructure/router.py%2026%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47334357%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22for%20every%20hop-by-hop%20extension%20we%20stop%20router%27s%20regular%20processing%3F%5Cr%5CnDoes%20it%20work%20with%20TracerouteExt%20now%3F%22%2C%20%22created_at%22%3A%20%222015-12-11T09%3A17%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22also%20do%20we%20have%20a%20rule%20that%20sibra%20ext%20is%20first%20%3F%22%2C%20%22created_at%22%3A%20%222015-12-11T09%3A19%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22No%2C%20for%20every%20hop-by-hop%20extension%20that%20signals%20we%20should%20stop%20regular%20processing%2C%20we%20stop%20regular%20processing.%20Note%20that%20the%20code%20checks%20the%20return%20value%20of%20the%20handler.%5Cr%5Cn%5Cr%5CnWe%27ll%20need%20to%20make%20this%20more%20granular%20in%20future%2C%20but%20this%20works%20for%20now.%22%2C%20%22created_at%22%3A%20%222015-12-14T09%3A17%3A39Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20infrastructure/router.py%3AL252-262%22%7D%2C%20%22Pull%20aba8ff163866ec86db5bc129068de5a9136dd944%20lib/packet/ext/sibra.py%20109%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47334940%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22still%20policy%20with%20assertions%20and%20exceptions%20in%20unclear%20to%20me%2C%20so%20why%20here%20assert%20%3F%22%2C%20%22created_at%22%3A%20%222015-12-11T09%3A25%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22Good%20catch%2C%20this%20should%20be%20an%20exception.%20Fixing.%22%2C%20%22created_at%22%3A%20%222015-12-14T09%3A19%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/ext/sibra.py%3AL1-405%22%7D%2C%20%22Pull%20aba8ff163866ec86db5bc129068de5a9136dd944%20lib/packet/ext/sibra.py%2055%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/netsec-ethz/scion/pull/558%23discussion_r47335024%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22somehow%20I%20prefer%20binary%20representation%20for%20flags...%22%2C%20%22created_at%22%3A%20%222015-12-11T09%3A26%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/8159928%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/pszalach%22%7D%7D%2C%20%7B%22body%22%3A%20%22We%20can%20do%20that%2C%20sure%2C%20It%27s%20just%20tricky%20counting%20the%200%27s.%20Changing.%22%2C%20%22created_at%22%3A%20%222015-12-14T09%3A22%3A12Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6919822%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kormat%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20lib/packet/ext/sibra.py%3AL1-405%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/558#issuecomment-163319295'>General Comment</a></b>
- <a href='https://github.com/kormat'></a> @pszalach
- <a href='https://github.com/kormat'></a> @reischuk
- <a href='https://github.com/kormat'></a> Ready for review now.
- <a href='https://github.com/pszalach'></a> the rest lgtm
- <a href='https://github.com/kormat'></a> PR updated.
- [ ] <a href='#crh-comment-Pull aba8ff163866ec86db5bc129068de5a9136dd944 infrastructure/router.py 26'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/558#discussion_r47334357'>File: infrastructure/router.py:L252-262</a></b>
- <a href='https://github.com/pszalach'></a> for every hop-by-hop extension we stop router's regular processing?
  Does it work with TracerouteExt now?
- <a href='https://github.com/pszalach'></a> also do we have a rule that sibra ext is first ?
- <a href='https://github.com/kormat'></a> No, for every hop-by-hop extension that signals we should stop regular processing, we stop regular processing. Note that the code checks the return value of the handler.
  We'll need to make this more granular in future, but this works for now.
- [ ] <a href='#crh-comment-Pull aba8ff163866ec86db5bc129068de5a9136dd944 lib/packet/ext/sibra.py 109'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/558#discussion_r47334940'>File: lib/packet/ext/sibra.py:L1-405</a></b>
- <a href='https://github.com/pszalach'></a> still policy with assertions and exceptions in unclear to me, so why here assert ?
- <a href='https://github.com/kormat'></a> Good catch, this should be an exception. Fixing.
- [ ] <a href='#crh-comment-Pull aba8ff163866ec86db5bc129068de5a9136dd944 lib/packet/ext/sibra.py 55'></a> &nbsp;<b><a href='https://github.com/netsec-ethz/scion/pull/558#discussion_r47335024'>File: lib/packet/ext/sibra.py:L1-405</a></b>
- <a href='https://github.com/pszalach'></a> somehow I prefer binary representation for flags...
- <a href='https://github.com/kormat'></a> We can do that, sure, It's just tricky counting the 0's. Changing.

<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/558?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/netsec-ethz/scion/pull/558?approve=1'></a>&nbsp;<a href='https://github.com/netsec-ethz/scion/pull/558'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: infrastructure/router.py [hunk 1]
```diff
@@ -250,9 +252,10 @@ def handle_extensions(self, spkt, pre_routing_phase):
                 logging.warning("No handler for extension type %u",
                                 ext_hdr.EXT_TYPE)
                 continue
-            handler(spkt=spkt, ext=ext_hdr, conf=self.config,
-                    topo=self.topology, iface=self.interface)
-        return False
+            if handler(spkt=spkt, ext=ext_hdr, conf=self.config,
+                       topo=self.topology, iface=self.interface):
+                handled = True
+        return handled
```
Comment (pszal) [inline infrastructure/router.py:258]: for every hop-by-hop extension we stop router's regular processing?
Does it work with TracerouteExt now?

Comment (pszal) [inline infrastructure/router.py:258]: also do we have a rule that sibra ext is first ?

File: lib/packet/ext/sibra.py [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+            for plen in path_lens:
+                if not plen:
+                    break
+                self.active_blocks.append(self._parse_block(data, plen))
+        if self.req:
+            if self.accepted:
+                self.req_block = self._parse_block(data, self.total_hops)
+            else:
+                self.req_block = self._parse_offers_block(data)
+        assert len(data) == 0
```
Comment (pszal) [inline lib/packet/ext/sibra.py]: still policy with assertions and exceptions in unclear to me, so why here assert ?

File: lib/packet/ext/sibra.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+          Set if packet is travelling src->dest.
+      - version (2b): SIBRA version, to be used as (SCION ver, SIBRA ver).
+    """
+    NAME = "SibraExt"
+    EXT_TYPE = HopByHopType.SIBRA
+    STEADY_ID_LEN = 8
+    EPHEMERAL_ID_LEN = 8
+    MIN_LEN = HopByHopExtension.SUBHDR_LEN
+    SIBRA_VERSION = 0
+    FLAG_PATH_SETUP = 2**7
```
Comment (pszal) [inline lib/packet/ext/sibra.py]: somehow I prefer binary representation for flags...

File: infrastructure/router.py (hunk 1 above)
Comment (kormat) [inline infrastructure/router.py:258]: No, for every hop-by-hop extension that signals we should stop regular processing, we stop regular processing. Note that the code checks the return value of the handler.

We'll need to make this more granular in future, but this works for now.

File: lib/packet/ext/sibra.py (hunk 2 above)
Comment (kormat) [inline lib/packet/ext/sibra.py]: Good catch, this should be an exception. Fixing.

File: lib/packet/ext/sibra.py (hunk 3 above)
Comment (kormat) [inline lib/packet/ext/sibra.py]: We can do that, sure, It's just tricky counting the 0's. Changing.

**Pull request merged**